src/app/api/toy-example/README.md

```diff
@@ -0,0 +1,398 @@
+4. `PATCH /api/students/[id]` to update information for a specific student with the given `id`.
+
+As the developer of the API routes, you get to make the contract between the client and the server. In other words, you're telling the client "when url `X` is entered with HTTP verb `Y`, the server will perform action `Z`". Therefore, it's equally valid to design the abovementioned example with:
+
+1. `GET /api/getAllStudents`,
+2. `GET /api/getStudentByID`,
+3. `POST /api/createStudent`,
+4. `PATCH /api/updateStudentByID`.
+
+I opt-in for the first approach whenever possible; the segments of the API route specifies the collections that they interact with and the HTTP verb dictates what action to perform on these collections. I think this approach is more maintable and intuitive to end-users.
+
+## A Toy Example
+
+I will now guide you through the process of designing an API route for users to create a new account with an application.
+
+I've also included the source code and comments within the directory. We will be using TypeScript, [Zod](https://zod.dev/), and [Next.js 13](https://nextjs.org/docs). I will assume that you have basic understanding of TypeScript and React.
+
+### Specifying The Contract
+
+We will first determine the input and output specifications when a new user sign up. In practice, the constraints is context-dependent.
+
+In our example, we will define the following contraints:
+
+1. Users sign up with a valid `email` and `password`,
+2. An error is returned if the chosen `email` already exists,
+3. The password needs to be at least 5 characters long,
+
+The request body will contain the new user's `email` and `password`, and the response body will contain a code specifying the result of the request and a message to explain the code.
+
+We will use the specifications to write schemas with `Zod` to validate the input and output.
+
+```ts
+// route.schema.ts
+
+import { z } from "zod";
+
+export const SignInRequest = z.object({
+  email: z.string().email({ message: "Username must be an email" }),
+  password: z
+    .string()
+    .min(5, { message: "Password must be at least 5 characters" }),
+});
+
+export const SignInResponse = z.discriminatedUnion("code", [
+  z.object({
+    code: z.literal("SUCCESS"),
+    message: z.literal("The new user was successfully registered"),
+  }),
+
+  z.object({
+    code: z.literal("EMAIL_EXIST"),
+    message: z.literal("The email existed already"),
+  }),
+
+  z.object({
+    code: z.literal("INVALID_EMAIL"),
+    message: z.literal("The email is invalid"),
+  }),
+
+  z.object({
+    code: z.literal("INVALID_PASSWORD"),
+    message: z.literal("The password is invalid"),
+  }),
+
+  z.object({
+    code: z.literal("UNKNOWN"),
+    data: z.any(),
+  }),
+]);
+```
+
+I hope the `SignInRequest` is intuitive. The `SignInResponse` is slightly more complex, but I will explain the motivation. From the specifications, we can determine 4 outcomes when a new user sign up:
```

Comment:
Sometimes folks get tripped up on what an "object" is if they're coming from a language other than JS. Might be worth calling out that an object is a "bundle of variables" and similar to a struct in C/C++